src/Dialect/ONNX/ONNXOps/NN/Attention.cpp

```diff
@@ -0,0 +1,201 @@
+      shape4D[3] = shape4D[3].floorDiv(shape4D[1]);
+    }
+    return shape4D;
+  };
+
+  DimsExpr qShape4D = normalizeInputTo4D(qShape, qNumHeads);
+  DimsExpr kShape4D = normalizeInputTo4D(kShape, kvNumHeads);
+  DimsExpr vShape4D = normalizeInputTo4D(vShape, kvNumHeads);
+
+  DimsExpr outputDims = qShape;
+  if (rank == 4) {
+    outputDims[3] = vShape4D[3];
+  } else /*if (rank == 3)*/ {
+    outputDims[2] = qShape4D[1] * vShape4D[3];
+  }
+  setOutputDims(outputDims, 0);
+
+  if (!hasShapeAndRank(attentionOp.getPastKey()) ||
+      !hasShapeAndRank(attentionOp.getPastValue()))
+    return success();
+
+  DimsExpr pastKShape;
+  createIE->getShapeAsDims(attentionOp.getPastKey(), pastKShape);
+  DimsExpr pastVShape;
+  createIE->getShapeAsDims(attentionOp.getPastValue(), pastVShape);
+
+  if (pastKShape.size() != 4 || pastVShape.size() != 4)
+    return failure();
+
+  auto totalSeqLen = pastKShape[2] + kShape4D[2];
+
+  DimsExpr presentKeyDims = kShape4D;
+  presentKeyDims[2] = totalSeqLen;
+  setOutputDims(presentKeyDims, 1);
+
+  DimsExpr presentValueDims = vShape4D;
+  presentValueDims[2] = totalSeqLen;
+  setOutputDims(presentValueDims, 2);
+
+  DimsExpr qkOutputDims = qShape4D;
+  qkOutputDims[3] = totalSeqLen;
+  setOutputDims(presentValueDims, 3);
+
+  return success();
+}
+
+} // namespace onnx_mlir
+
+//===----------------------------------------------------------------------===//
+// Verify
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXAttentionOp::verify() {
+  // If presentK and presentV are outputs, then we must pass pastK and pastV as
+  // inputs
+  Value presentK = this->getResult(1);
+  Value presentV = this->getResult(2);
+  if (!isNoneValue(presentK) || !isNoneValue(presentV)) {
+    Value pastK = this->getOperand(4);
+    Value pastV = this->getOperand(5);
+    if (isNoneValue(pastK) || isNoneValue(pastV))
+      return emitOpError("inputs 'pastK' and 'pastV' are needed for outputs "
+                         "'presentK' and 'presentV'");
+  }
+
+  ONNXAttentionOpAdaptor adaptor(*this);
+
+  Value q = adaptor.getQ();
+  if (!hasShapeAndRank(q))
+    return success(); // Won't be able to do any more checking at this stage.
+
+  auto qType = mlir::cast<ShapedType>(q.getType());
+  int64_t qRank = qType.getShape().size();
+  if (qRank != 3 && qRank != 4)
+    return onnx_mlir::Diagnostic::emitOperandHasUnexpectedRankError(
+        *this->getOperation(), q, qRank, "3 or 4");
+
+  // check q_num_heads is present for 3D input
+  auto qNumHeads = adaptor.getQNumHeads();
+  if (qRank == 3 && !qNumHeads)
+    return emitOpError("attribute 'q_num_heads' must be provided when input "
+                       "'q' is a 3D tensor.");
+
+  Value k = adaptor.getK();
+  Value v = adaptor.getV();
+  if (!hasShapeAndRank(k) || !hasShapeAndRank(v))
+    return success(); // Won't be able to do any more checking at this stage.
+
+  auto kType = mlir::cast<ShapedType>(k.getType());
```

Comment:
Nit: You are already using the mlir namespace, but this is mostly a style preference
```suggestion
  auto kType = cast<ShapedType>(k.getType());
```